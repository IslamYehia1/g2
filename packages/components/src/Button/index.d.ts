export * from '../../types/components/Button';
