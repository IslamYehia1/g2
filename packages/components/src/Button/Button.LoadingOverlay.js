import { styled } from '@g2/css';
import React from 'react';

import { Flex } from '../Flex';
import { Spinner } from '../Spinner';

const LoadingOverlayView = styled(Flex)`
	pointer-events: none;
	position: absolute;
	top: 0;
	left: 0;
	bottom: 0;
	right: 0;
`;

export function LoadingOverlay({ isLoading = false }) {
	if (!isLoading) return null;

	return (
		<LoadingOverlayView aria-hidden="true" justify="center">
			<Spinner color={'currentColor'} />
		</LoadingOverlayView>
	);
}

export default LoadingOverlay;
