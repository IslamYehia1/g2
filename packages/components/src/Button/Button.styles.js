import { css, styled } from '@g2/css';

import { FlexItem } from '../Flex';

const baseStyles = ({ isDestructive, theme }) => {
	const { config } = theme;
	const { isDark } = config;
	const buttonTransition = `all ${config.buttonTransitionDuration}
    ${config.buttonTransitionTimingFunction}`;

	return css`
		align-items: center;
		appearance: none;
		background-color: ${config.buttonBackgroundColor};
		border-radius: ${config.buttonBorderRadius};
		border-color: transparent;
		border-style: solid;
		border-width: ${config.buttonBorderWidth};
		box-shadow: ${config.buttonBoxShadow};
		color: ${config.colorText};
		cursor: pointer;
		display: inline-flex;
		font-weight: ${config.buttonFontWeight};
		height: ${config.buttonHeight};
		line-height: ${config.buttonHeight};
		outline: none;
		padding-left: ${config.buttonPaddingX};
		padding-right: ${config.buttonPaddingX};
		position: relative;
		transform: ${config.buttonTransform};
		text-decoration: none;
		user-select: none;
		width: auto;

		&:hover,
		&:focus {
			background-color: ${config.buttonBackgroundColorHover};
		}

		&:active {
			background-color: ${config.buttonBackgroundColorActive};
			transform: ${config.buttonTransformActive};
			transform-origin: center center;
		}

		&:hover,
		&:active,
		&:focus {
			transition: ${buttonTransition};
		}

		&[disabled],
		&[aria-disabled='true'] {
			cursor: auto;
			opacity: 0.5;
		}

		svg {
			display: block;
		}

		${isDark &&
		css`
			background-color: ${config.buttonBackgroundColorDark};
			color: ${config.colorTextDark};

			&:hover,
			&:focus {
				background-color: ${config.buttonBackgroundColorHoverDark};
			}

			&:active {
				background-color: ${config.buttonBackgroundColorActiveDark};
			}

			${isDestructive &&
			css`
				color: ${config.colorDestructive};
			`}
		`}
	`;
};

const blockStyles = ({ isBlock }) => {
	if (!isBlock) return;

	return css`
		display: flex;
		width: 100%;

		&:active {
			transform: scale(1);
		}
	`;
};

const roundedStyles = ({ isRounded, theme }) => {
	if (!isRounded) return '';

	const { config } = theme;

	return css`
		border-radius: ${config.buttonBorderRadiusRound};
	`;
};

const sizeStyles = ({ size, theme }) => {
	const { config } = theme;

	switch (size) {
		case 'large':
			return css`
				height: ${config.buttonHeightLarge};
				line-height: ${config.buttonHeightLarge};
			`;

		case 'small':
			return css`
				height: ${config.buttonHeightSmall};
				line-height: ${config.buttonHeightSmall};
			`;

		default:
			return '';
	}
};

const primaryStyles = ({ isDestructive, isOutline, theme }) => {
	const { config } = theme;

	return css`
		background-color: ${config.buttonBackgroundColorPrimary};
		color: ${config.buttonTextColorPrimary};

		&:hover,
		&:focus {
			background-color: ${config.buttonBackgroundColorPrimaryHover};
		}

		&:focus {
			border-color: ${config.buttonBorderColorFocus};
			box-shadow: ${config.buttonBoxShadowFocus};
		}

		&:active {
			background-color: ${config.buttonBackgroundColorPrimaryActive};
		}

		${isDestructive &&
		css`
			background-color: ${config.colorDestructive};

			&:hover,
			&:focus {
				background-color: ${config.colorDestructiveHover};
			}

			&:focus {
				border-color: ${config.buttonBorderColorFocus};
				box-shadow: ${config.buttonBoxShadowDestructiveFocus};
			}

			&:active {
				background-color: ${config.colorDestructiveActive};
			}
		`}

		${isOutline &&
		css`
			background-color: transparent;
			border-color: ${config.buttonBackgroundColorPrimary};
			color: ${config.buttonBackgroundColorPrimary};

			&:hover,
			&:active,
			&:focus {
				color: ${config.buttonTextColorPrimary};
			}
		`}
	`;
};

const secondaryStyles = ({ isDestructive, isOutline, theme }) => {
	if (!isOutline) return '';

	const { config } = theme;
	const { isDark } = config;

	return css`
		background-color: transparent;
		border-color: ${config.buttonBorderColorOutline};

		${isDark &&
		`
			border-color: ${config.buttonBorderColorOutlineDark};
		`}

		&:hover,
		&:active,
		&:focus {
			border-color: transparent;
		}

		${isDestructive &&
		`
			color: ${config.colorDestructive};
			border-color: ${config.colorDestructive};
		`}
	`;
};

const tertiaryStyles = () => {
	return css`
		background-color: transparent;
	`;
};

const linkStyles = ({ isDestructive, theme }) => {
	const { config } = theme;

	return css`
		background: none;
		border-color: transparent;
		color: ${config.buttonBackgroundColorPrimary};

		${isDestructive &&
		css`
			color: ${config.colorDestructive};
		`}
	`;
};

const plainLinkStyles = (props) => css`
	${linkStyles(props)};

	padding-left: 0;
	padding-right: 0;

	&:hover,
	&:active,
	&:focus {
		background-color: transparent;
		text-decoration: underline;
	}
`;

const variantStyles = ({ variant, ...props }) => {
	switch (variant) {
		case 'primary':
			return primaryStyles(props);
		case 'secondary':
			return secondaryStyles(props);
		case 'tertiary':
			return tertiaryStyles(props);
		case 'link':
			return linkStyles(props);
		case 'plain-link':
			return plainLinkStyles(props);
		default:
			return '';
	}
};

export const ButtonView = styled.BaseView`
	${baseStyles};
	${blockStyles};
	${variantStyles};
	${roundedStyles};
	${sizeStyles};
`;

const contentStyles = ({ theme }) => {
	const { config } = theme;

	return css`
		line-height: ${config.buttonContentLineHeight};
		min-height: ${config.buttonLineHeight};
		white-space: nowrap;
	`;
};

const loadingStyles = ({ isLoading }) => css`
	opacity: 1;

	${isLoading &&
	`
    opacity: 0;
`}
`;

export const ContentView = styled(FlexItem)`
	${contentStyles};
	${loadingStyles};
`;

const prefixSuffixStyles = ({ theme }) => {
	const { config } = theme;

	return css`
		line-height: ${config.inputLineHeight};

		svg {
			display: block;
			user-select: none;
		}
	`;
};

export const PrefixSuffixView = styled(FlexItem)`
	${prefixSuffixStyles};
	${loadingStyles};
`;

const caretWrapperStyles = ({ theme }) => {
	const { space } = theme?.utils;
	return css`
		margin-right: ${space(-1.5)};
	`;
};

export const CaretWrapperView = styled(FlexItem)`
	${caretWrapperStyles};
	${loadingStyles};
`;
