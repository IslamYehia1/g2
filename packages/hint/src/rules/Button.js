export default {
	rules: {
		'button-siblings-layout-wrap': {
			create: (context) => {
				return (node, { getComponentName }) => {
					const isButton = getComponentName(node) === 'Button';
					if (!isButton) return;

					const sibling = node.nextElementSibling;
					const isNextButton = getComponentName(sibling) === 'Button';
					if (isNextButton) {
						context.report({
							title: 'Button / Siblings: Layout wrap',
							message:
								'Buttons should not be next to each other. Wrap them with a Layout container, such as Flex, HStack, or VStack.',
						});
					}
				};
			},
		},
	},
};
