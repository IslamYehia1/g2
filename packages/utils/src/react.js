import React from 'react';
import { is } from './is';

export function getValidChildren(children) {
	return React.Children.toArray(children).filter((child) =>
		React.isValidElement(child),
	);
}

export function assignRef(ref, value) {
	if (ref == null) return;

	if (is.function(ref)) {
		ref(value);
		return;
	}

	try {
		//@ts-ignore
		ref.current = value;
	} catch (error) {
		throw new Error(`Cannot assign value '${value}' to ref '${ref}'`);
	}
}

export function mergeRefs(...refs) {
	return (value) => {
		refs.forEach((ref) => assignRef(ref, value));
	};
}
