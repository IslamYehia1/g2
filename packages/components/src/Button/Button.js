import { connect } from '@wp-g2/provider';
import React from 'react';
import { FiChevronDown } from 'react-icons/fi';
import { Button as BaseButton } from 'reakit/Button';

import { useControlGroupContext } from '../ControlGroup';
import { Elevation } from '../Elevation';
import { Flex } from '../Flex';
import { Icon } from '../Icon';
import LoadingOverlay from './Button.LoadingOverlay';
import {
	ButtonLinkView,
	ButtonView,
	CaretWrapperView,
	ContentView,
	PrefixSuffixView,
} from './Button.styles';

function Button({
	children,
	elevation = 0,
	elevationActive,
	elevationFocus,
	elevationHover,
	forwardedRef,
	gap = 2,
	hasCaret = false,
	href,
	isBlock = false,
	isDestructive = false,
	isLoading = false,
	isNarrow = false,
	isRounded = false,
	isSubtle = false,
	justify = 'center',
	prefix,
	size = 'medium',
	suffix,
	variant = 'secondary',
	...props
}) {
	const as = href ? ButtonLinkView : ButtonView;
	const { isFirst, isLast, isMiddle } = useControlGroupContext();

	return (
		<BaseButton
			aria-busy={isLoading}
			as={as}
			href={href}
			isBlock={isBlock}
			isDestructive={isDestructive}
			isFirst={isFirst}
			isLast={isLast}
			isMiddle={isMiddle}
			isRounded={isRounded}
			isSubtle={isSubtle}
			ref={forwardedRef}
			size={size}
			variant={variant}
			{...props}
		>
			<LoadingOverlay isLoading={isLoading} />
			<Flex as="span" gap={gap} justify={justify} sx={{ width: '100%' }}>
				{prefix && (
					<PrefixSuffixView as="span" isLoading={isLoading}>
						{prefix}
					</PrefixSuffixView>
				)}
				<ContentView as="span" isLoading={isLoading}>
					{children}
				</ContentView>
				{suffix && (
					<PrefixSuffixView as="span" isLoading={isLoading}>
						{suffix}
					</PrefixSuffixView>
				)}
				{hasCaret && (
					<CaretWrapperView as="span" isLoading={isLoading}>
						<Icon icon={<FiChevronDown />} size={16} />
					</CaretWrapperView>
				)}
			</Flex>
			<Elevation
				active={elevationActive}
				as="span"
				focus={elevationFocus}
				hover={elevationHover}
				offset={-1}
				value={elevation}
			/>
		</BaseButton>
	);
}

export default connect(Button, 'Button');
