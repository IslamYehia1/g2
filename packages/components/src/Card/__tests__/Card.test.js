import { render } from '@testing-library/react';
import React from 'react';

import { Button } from '../../Button';
import { Card, CardBody, CardFooter, CardHeader } from '../index';

describe('props', () => {
	test('should render correctly', () => {
		const { container } = render(
			<Card>
				<CardHeader title="Olaf" />
				<CardBody>Some people are worth melting for.</CardBody>
				<CardFooter>
					<Button variant="primary">Agree</Button>
				</CardFooter>
			</Card>,
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	test('should render elevation', () => {
		const { container } = render(<Card elevation={5} />);
		expect(container.firstChild).toMatchSnapshot();
	});
});
