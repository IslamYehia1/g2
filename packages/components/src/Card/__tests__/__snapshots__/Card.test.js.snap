// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props should render CardInnerBody 1`] = `
<div
  class="wp-components-surface wp-components-card css-azv6f"
  data-g2-c16t="true"
  data-g2-component="Card"
>
  <div
    class="wp-components-view css-j1yu9m"
    data-g2-c16t="true"
    data-g2-component="CardContent"
  >
    <div
      class="wp-components-view wp-components-flex wp-components-card-header css-tycr7a"
      data-g2-c16t="true"
      data-g2-component="CardHeader"
      title="Olaf"
    />
    <div
      class="wp-components-card-inner-body css-1jwn1my"
      data-g2-c16t="true"
      data-g2-component="CardInnerBody"
    >
      Some people are worth melting for.
    </div>
    <div
      class="wp-components-view wp-components-flex wp-components-card-footer css-xk6v0j"
      data-g2-c16t="true"
      data-g2-component="CardFooter"
    >
      <div
        class="wp-components-flex-item css-1suoifq"
        data-g2-c16t="true"
        data-g2-component="FlexItem"
      >
        <button
          aria-busy="false"
          class="wp-components-button css-1jo4blp"
          data-destructive="false"
          data-g2-c16t="true"
          data-g2-component="Button"
          data-icon="false"
          type="button"
        >
          <span
            class="wp-components-view wp-components-flex css-2r7rsp"
            data-g2-c16t="true"
            data-g2-component="Flex"
          >
            <span
              class="wp-components-flex-item wp-components-flex-block css-13400lq"
              data-g2-c16t="true"
              data-g2-component="ButtonContent"
            >
              Agree
            </span>
          </span>
          <span
            class="wp-components-elevation css-17x6dun"
            data-g2-c16t="true"
            data-g2-component="ButtonElevation"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="wp-components-elevation css-18hqkt0"
    data-g2-c16t="true"
    data-g2-component="CardElevation"
  />
  <div
    class="wp-components-elevation css-1873flw"
    data-g2-c16t="true"
    data-g2-component="CardElevation"
  />
</div>
`;

exports[`props should render correctly 1`] = `
<div
  class="wp-components-surface wp-components-card css-azv6f"
  data-g2-c16t="true"
  data-g2-component="Card"
>
  <div
    class="wp-components-view css-j1yu9m"
    data-g2-c16t="true"
    data-g2-component="CardContent"
  >
    <div
      class="wp-components-view wp-components-flex wp-components-card-header css-tycr7a"
      data-g2-c16t="true"
      data-g2-component="CardHeader"
      title="Olaf"
    />
    <div
      class="wp-components-scrollable wp-components-card-body css-3pyc4i"
      data-g2-c16t="true"
      data-g2-component="CardBody"
    >
      <div
        class="css-1qs21wj"
      >
        Some people are worth melting for.
      </div>
    </div>
    <div
      class="wp-components-view wp-components-flex wp-components-card-footer css-xk6v0j"
      data-g2-c16t="true"
      data-g2-component="CardFooter"
    >
      <div
        class="wp-components-flex-item css-1suoifq"
        data-g2-c16t="true"
        data-g2-component="FlexItem"
      >
        <button
          aria-busy="false"
          class="wp-components-button css-1jo4blp"
          data-destructive="false"
          data-g2-c16t="true"
          data-g2-component="Button"
          data-icon="false"
          type="button"
        >
          <span
            class="wp-components-view wp-components-flex css-2r7rsp"
            data-g2-c16t="true"
            data-g2-component="Flex"
          >
            <span
              class="wp-components-flex-item wp-components-flex-block css-13400lq"
              data-g2-c16t="true"
              data-g2-component="ButtonContent"
            >
              Agree
            </span>
          </span>
          <span
            class="wp-components-elevation css-17x6dun"
            data-g2-c16t="true"
            data-g2-component="ButtonElevation"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="wp-components-elevation css-18hqkt0"
    data-g2-c16t="true"
    data-g2-component="CardElevation"
  />
  <div
    class="wp-components-elevation css-1873flw"
    data-g2-c16t="true"
    data-g2-component="CardElevation"
  />
</div>
`;

exports[`props should render elevation 1`] = `
<div
  class="wp-components-surface wp-components-card css-azv6f"
  data-g2-c16t="true"
  data-g2-component="Card"
>
  <div
    class="wp-components-view css-j1yu9m"
    data-g2-c16t="true"
    data-g2-component="CardContent"
  />
  <div
    class="wp-components-elevation css-18hqkt0"
    data-g2-c16t="true"
    data-g2-component="CardElevation"
  />
  <div
    class="wp-components-elevation css-zytxex"
    data-g2-c16t="true"
    data-g2-component="CardElevation"
  />
</div>
`;

exports[`props should render no elevation 1`] = `
<div
  class="wp-components-surface wp-components-card css-azv6f"
  data-g2-c16t="true"
  data-g2-component="Card"
>
  <div
    class="wp-components-view css-j1yu9m"
    data-g2-c16t="true"
    data-g2-component="CardContent"
  />
  <div
    class="wp-components-elevation css-16j6c8z"
    data-g2-c16t="true"
    data-g2-component="CardElevation"
  />
  <div
    class="wp-components-elevation css-12htjgm"
    data-g2-c16t="true"
    data-g2-component="CardElevation"
  />
</div>
`;
